app/lib/import_export.rb

```diff
@@ -11,497 +12,186 @@ module ImportExport # rubocop:todo Metrics/ModuleLength
+        csvs = {}
+        packages = {}
+        data_types.each do |data_type|
+          # Create CSV with column headers
+          package = CSV.generate(headers: true) do |csv|
+            csv << field_data.dig(data_type, :headers)
+            csvs[data_type] = csv
+          end
+          packages[data_type] = package
+        end
+      end
+
+      # 2) Get export data in batches to decrease size of export data hash maintained in memory
+      exported_data = get_export_data(batch_group.order(:id), config[:data])
+      data_types.each do |data_type|
+        exported_data[data_type]&.each do |record|
+          csvs[data_type] << field_data[data_type][:checked].map { |field| record[field] }
+        end
+      end
+
+      # 3) Save files to db at the end of each outer batch
+      next unless ((inner_batch_index + 1) % (outer_batch_size / inner_batch_size)).zero? || inner_batch_index == total_inner_batches - 1
+
+      outer_batch_index = inner_batch_index * inner_batch_size / outer_batch_size
+      data_types.each do |data_type|
+        file = { filename: build_export_filename(config, data_type, outer_batch_index, false), content: Base64.encode64(packages[data_type]) }
+        files << save_file(config, file)
+      end
+    end
+
+    files
+  end
+
+  # Creates a list of excel files from exported data
+  def xlsx_export(config, patients, outer_batch_size, inner_batch_size)
+    files = []
+    separate_files = config[:separate_files].present?
+
+    # Determine selected data types for export
+    data_types = CUSTOM_EXPORT_OPTIONS.keys.select { |data_type| config.dig(:data, data_type, :checked).present? }
+
+    # Get all of the field data based on the config
+    field_data = get_field_data(config)
+
+    # Declare variables in scope outside of batch loop
+    workbooks = nil if separate_files
+    workbook = nil unless separate_files
+    sheets = nil
+    last_row_nums = nil
+    total_inner_batches = (patients.size / inner_batch_size.to_f).ceil
+
+    # NOTE: in_batches appears to NOT sort within batches, so explicit ordering on ID is also done deeper down.
+    # The reorder('') here allows this ordering done later on to work properly.
+    patients.reorder('').in_batches(of: inner_batch_size).each_with_index do |batch_group, inner_batch_index|
+      # 1) Create file(s) for new batch
+      if (inner_batch_index % (outer_batch_size / inner_batch_size)).zero?
+        # One file for all data types, each data type in a different tab
+        workbooks = {} if separate_files
+        workbook = FastExcel.open(constant_memory: true) unless separate_files
+        sheets = {}
+        last_row_nums = {}
+        data_types.each do |data_type|
+          workbook = FastExcel.open(constant_memory: true) if separate_files
+          worksheet = workbook.add_worksheet(config.dig(:data, data_type, :tab) || CUSTOM_EXPORT_OPTIONS.dig(data_type, :label))
+          worksheet.auto_width = true
```

Comment:
Since we're calling `write_string` directly for most of the writes instead of any of the helper methods such as `append_row` or `write_value`, the `add_text_width` method (which I'm assuming makes it slow) doesn't get called: https://github.com/Paxa/fast_excel/blob/8fafe6f20cf51c683314a589f1597113d8704b0a/lib/fast_excel.rb#L445

It only gets called for the header row. I thought this would be a decent compromise between making the columns at least somewhat readable and formatted while still maintaining performance.

Reply:
That's fair - I'm fine with that.